### DISCUSSION TITLE
[Solved] Specifying Selectors

### DISCUSSION THREAD

**Discussion by p-avital** [eclipse-zenoh/roadmap#25] [Ideas and new features]
Much like Key Expressions, although a bit less core to Zenoh, Selectors are an important concept in Zenoh. We should specify them properly and address possible breaking changes we could make with `0.6.0`.

## Opacity
Currently, selectors are wide open. This gives us very little flexibility in their implementation. Selectors should become opaque with proper accessors to stop leaking their implementation details.

## Renaming
Currently, selectors are composed of a `key_selector` and a `value_selector`. We already have a name for something that selects keys: "key expression", so we should rename `key_selector` to `key_expr`, which would remove potential misunderstandings that "key selectors" are in some way different from "key expressions".

As for `value_selector`: it does hold some sense since part of its role is filtering according to value, but it's also way more generic than this considering queryables are free to interpret them however they like. For the rest of this discussion, I'll refer to it as the "complement" portion of a selector.

## Construction
Currently, complements must always hold their starting `?`. This just implies that anything that processes selector complements has to first discard the `?` before doing anything useful.

My biggest issue with this however is that users are expected to write `Selector::from(key_expr).with_value_selector("?some=complement(...)")` when constructing a complemented selector from a `KeyExpr`. There's no good reason to place such a trap for users.

## Complement format
Currently, there is no proper standard for serializing and deserializing complements. Our current public stance is that they should be of the shape `?x=4&y<3(arg1=value1;arg2=value2)[x;fragment]`.

Note that the filter and fragment sections can only work for self-describing values, and are somewhat under-specified. On top of that, these two sections are non-trivial to implement properly/consistently for users that would write their own queryables.

Meanwhile, "properties" could use a simple rename into "arguments", since they're really just RPC arguments.

I propose that we change the format such that the complement becomes `<arguments>(<filter>)[<projection>]` [(here's a regex proposal for clarity)](https://regex101.com/r/lON2J3/2):
* A unified self-describing format should be defined for arguments, such that even if a query is received with unknown arguments, at least their parsing is doable by any queryable.
* Filters get access to parentheses, and become allowed to be a more complex language. A possible nicety would be to support named predicates that queryables could evaluate locally. In the regex, I added the ability to use `!` as a "greediness" qualifier, which could let the user chose whether or not they want the values for which the filter couldn't be fully evaluated.
* The projection part remains to be properly defined, but some ideas entertained internally have some rather complex implementation logic that we should spare queryable-implementers by providing appropriate helpers.

Further discussion is needed for the exact format of each complement section, but I believe a well defined complement format would help with interoperability of differently-typed queryables.

**Comment by p-avital** [comment]
Of note, @Mallets PM'd me about graphql. We should study whether it's a good fit for Zenoh instead of (or as part of) the complement format

**Comment by p-avital** [comment]
This was further discussed with @Mallets, @kydos and @sreeja. Here's the current perspective on the right side of selectors:
* The name still lacks consensus, proposed names are `parameters`, `arguments` and `complements`.
* As far as `zenoh` is concerned, this right side is just a string, without an explicitly defined format. The user is free to place anything in it, much like they'd be free to place anything in a message's `Value`.
* In the future, after it has been properly discussed, `zenoh-ext` will provide facilities to help with parsing from a "encouraged" format, which should be designed to allow DSLs such as SQL to be written in.
  * These facilities should also help the user handle `filters` and `projections` properly, as the distributed nature of `zenoh` makes filtering much harder than it seems, and the key-expression compatibility of projections that we currently envisions can be very hard to implement properly.

**Comment by p-avital** [comment]
# Final decisions:
The `value_selector` format is changing to become more simpler and more familiar, while giving the Zenoh team more freedom to integrate it to Zenoh.

## Format
A `Selector` remains formatted as `<key/expression>?<value_selector>`, however the `value_selector` section changes to the same format as a URL query section. It becomes abstracted as a list of key-value pairs, formatted such that:
- key-value pairs are separated by `&`.
- each pair follows the `(?<key>[^=]*)(=(?<value>[^&]*))?` regex: note that the value section is optional: `somekey` parses to `("somekey", "")`.
- both the `key` and `value` fields shall be URL-encoded to allow for character escapement.

## Conventions
Queryable implementers will be encouraged to follow the following conventions, which the Zenoh team shall abide:
- The Zenoh Team considers any key that doesn't start with an alphanumeric character reserved for its own uses. Zenoh may adapt its behavior to the presence and values associated with any key that doesn't start with an alphanumeric character.
- The Zenoh Team will, over time, specify the use of certain keys, associated with certain formats, for specific purposes. Queryable implementers are encouraged to use the specified fields (the Zenoh Team will provide utilities to facilitate proper use) when appropriate.
  - For example, queryables that maintain a history of values are encouraged to use the `_time` key
as a way to obtain the range of time the querier is interested in. Utilities to parse the value, expected to be a range expressed through the Zenoh Time DSL, will be provided to help compliant queryables.
  - The `_filter` and `_projection` keys are expected to be standardized at some point, with relevant utilities to ensure ease of implementation of compliant queryables.
- To avoid collisions of keys between queryables that may treat them with different semantics, implementers are encouraged to select a prefix for their own keys.

**Comment by Mallets** [comment]
Locking discussion. RFC for Selectors is available here: https://github.com/eclipse-zenoh/roadmap/tree/main/rfcs/ALL/Selectors
